library/src/meta/figure.rs

```diff
@@ -94,3 +163,110 @@ impl LocalName for FigureElem {
         }
     }
 }
+
+impl ReferenceInfo for FigureElem {
+    fn counter(&self, _: StyleChain) -> Option<Counter> {
+        self.determine_type().counter()
+    }
+
+    fn numbering(&self, styles: StyleChain) -> Option<Numbering> {
+        self.numbering(styles)
+    }
+
+    fn supplement(&self, styles: StyleChain) -> Option<Supplement> {
+        self.determine_type().supplement(self, styles)
+    }
+}
+
+/// The type of a figure
+///
+/// Priority list:
+/// 1. `counter` and `supplement` explicitly set
+/// 2. contains an image element
+/// 3. contains a raw element
+/// 4. contains a table element
+/// 5. could not determine content
+#[derive(Debug, Clone)]
+#[repr(u8)]
+pub enum FigureType {
+    /// A figure containing one (or more) images
+    Image(ImageElem),
+
+    /// A figure containing a table
+    Table(TableElem),
+
+    /// A figure containing a snippet of code
+    Raw(RawElem),
+
+    /// Could not determine the content of the figure.
+    /// Unless the figure has `counter` and `supplement` explicitly set,
+    /// this will be treated as an error.
+    Other,
+}
+
+impl FigureType {
+    /// Is this figure type unknown?
+    pub fn is_other(&self) -> bool {
+        matches!(self, Self::Other)
+    }
+
+    /// Gets the function of the element associated with this figure type.
+    pub fn func(&self) -> Option<ElemFunc> {
+        match self {
+            FigureType::Image(_) => Some(ImageElem::func()),
+            FigureType::Table(_) => Some(TableElem::func()),
+            FigureType::Raw(_) => Some(RawElem::func()),
+            FigureType::Other => None,
+        }
+    }
+
+    /// Gets the counter associated with this figure type.
+    pub fn counter(&self) -> Option<Counter> {
+        match self {
+            FigureType::Image(_) => {
+                Some(Counter::new(CounterKey::Str("figure_images".into())))
```

Comment:
I know!!!! I tried to implement this, but it was so f-ing ugly, I couldn't take it. I don't know enough about the internals of typsts to make that change in a satisfactory way. The way I tried doing it (it was 3 AM so I was running on fumes) was that I would have a hidden arguments in tables, images and raw that would have a "in_figure" boolean. But I couldn't figure out how to set this flag nicely. Additionally, it has one more problem: if a figure contains multiple images, or images in a table, etc. it would increase all counters when only increasing it once would be needed.

Reply:
I think this would need some change to the counting code itself. It should also search for figure of the given type. But it's definitely complicated. I can try taking a stab at this once the PR is otherwise finished.

Reply:
It could be made such that counters can have a "sub type" so that `counter(figure)` is value but also `counter(figure, image)`?

Reply:
This could even apply to things like the outline, rather than `outline(of: image)` for what is actually the outline of images in figures, it could be something like `outline(of: (figure, image))` something along those lines